x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -0,0 +1,418 @@
+    return (this.indexInitialization = new Promise(async (resolve) => {
+      // Check if required index template exists.
+      let indexTemplateExists = false;
+      try {
+        indexTemplateExists = await this.options.clusterClient.callAsInternalUser(
+          'indices.existsTemplate',
+          { name: SESSION_INDEX_TEMPLATE.name }
+        );
+      } catch (err) {
+        this.options.logger.error(
+          `Failed to check if session index template exists: ${err.message}`
+        );
+        throw err;
+      }
+
+      // Create index template if it doesn't exist.
+      if (indexTemplateExists) {
+        this.options.logger.debug('Session index template already exists.');
+      } else {
+        try {
+          await this.options.clusterClient.callAsInternalUser('indices.putTemplate', {
+            name: SESSION_INDEX_TEMPLATE.name,
+            body: SESSION_INDEX_TEMPLATE.template,
+          });
+          this.options.logger.debug('Successfully created session index template.');
+        } catch (err) {
+          this.options.logger.error(`Failed to create session index template: ${err.message}`);
+          throw err;
+        }
+      }
+
+      // Check if required index exists. We cannot be sure that automatic creation of indices is
+      // always enabled, so we create session index explicitly.
+      let indexExists = false;
+      try {
+        indexExists = await this.options.clusterClient.callAsInternalUser('indices.exists', {
+          index: SESSION_INDEX_NAME,
+        });
+      } catch (err) {
+        this.options.logger.error(`Failed to check if session index exists: ${err.message}`);
+        throw err;
+      }
+
+      // Create index if it doesn't exist.
+      if (indexExists) {
+        this.options.logger.debug('Session index already exists.');
+      } else {
+        try {
+          await this.options.clusterClient.callAsInternalUser('indices.create', {
+            index: SESSION_INDEX_NAME,
+          });
+          this.options.logger.debug('Successfully created session index.');
+        } catch (err) {
+          // There can be a race condition if index is created by another Kibana instance.
+          if (err?.body?.error?.type === 'resource_already_exists_exception') {
+            this.options.logger.debug('Session index already exists.');
+          } else {
+            this.options.logger.error(`Failed to create session index: ${err.message}`);
+            throw err;
+          }
+        }
+      }
+
+      // Notify any consumers that are awaiting on this promise and immediately reset it.
+      resolve();
+      this.indexInitialization = undefined;
+    }));
+  }
+
+  /**
+   * Trigger a removal of any outdated session values.
+   */
+  async cleanUp() {
+    const now = Date.now();
+
+    // Always try to delete sessions with expired lifespan (even if it's not configured right now).
+    const deleteQueries: object[] = [{ range: { lifespanExpiration: { lte: now } } }];
```

Comment:
No particular benefit, calculating `now` here felt a bit more predictable and related to the time when user request actually reaches Kibana. I'm not really sure when `now` would evaluate to a number for every of the sub-query. But I'm fine either way, do you know any benefit in moving this to ES instead?

Reply:
> do you know any benefit in moving this to ES instead?

No real benefits came to mind. My only thought was my systemic fear of JavaScript's date handling. We're relying on the native Date object which has always scared me. This is likely unfounded though, since we're getting the epoch milliseconds directly, and not transitioning from one date format to another.

Let's leave it the way you have it, I had no founded objections at all to your approach.